packages/kbn-docs-utils/src/api_docs/utils.ts

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { KibanaPlatformPlugin } from '@kbn/dev-utils';
+import {
+  AnchorLink,
+  ApiDeclaration,
+  ScopeApi,
+  TypeKind,
+  Lifecycle,
+  PluginApi,
+  ApiScope,
+} from './types';
+
+function capitalize(str: string): string {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+}
+
+export const camelToSnake = (str: string): string => str.replace(/([A-Z])/g, '_$1').toLowerCase();
+
+export const snakeToCamel = (str: string): string =>
+  str.replace(/([-_][a-z])/g, (group) => group.toUpperCase().replace('-', '').replace('_', ''));
+
+/**
+ * Returns the plugin that the file belongs to.
+ * @param path An absolute file path that can point to a file nested inside a plugin
+ * @param plugins A list of plugins to search through.
+ */
+export function getPluginForPath(
+  path: string,
+  plugins: KibanaPlatformPlugin[]
+): KibanaPlatformPlugin | undefined {
+  return plugins.find((plugin) => path.startsWith(plugin.directory));
+}
+
+/**
+ * Groups ApiDeclarations by typescript kind - classes, functions, enums, etc, so they
+ * can be displayed separately in the mdx files.
+ */
+export function groupPluginApi(declarations: ApiDeclaration[]): ScopeApi {
+  const scope = createEmptyScope();
+
+  declarations.forEach((declaration) => {
+    addApiDeclarationToScope(declaration, scope);
+  });
+
+  return scope;
+}
+
+/**
+ * If the file is at the top level, returns undefined, otherwise returns the
+ * name of the first nested folder in the plugin. For example a path of
+ * 'src/plugins/data/public/search_services/file.ts' would return 'search_service' while
+ * 'src/plugin/data/server/file.ts' would return undefined.
+ * @param path
+ */
+export function getServiceForPath(path: string): string | undefined {
```

Comment:
There's a bug with `getServiceForPath` if we have a path like `"src/plugins/embed/server/routes/public/foo/index.ts"`. There's only one usage of `getServiceForPath` https://github.com/elastic/kibana/pull/86232/files#diff-d3ce0084b5713e36ab23a452ffe50380efc2b94fda541b83b502ab752b25ecabR47 and the `scope` is known. Perhaps we should just pass the scope to `getServiceForPath` and use this to create the regex?

Reply:
There are two usages, and the one in `getPluginApiDocId` doesn't have scope available.  I added a test and refactored this. The code that has scope available is just part of that function now, and the other usage passes in the plugin directory.